src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  function deepClone(obj) {
+    return JSON.parse(JSON.stringify(obj));
+}
+
+let currentState = deepClone(state);
+const stateHistory = [];
+
+for (const action of actions) {
+    switch (action.type) {
+        case 'clear':
+            currentState = {};
+            break;
+        case 'addProperties':
+            currentState = {
+                ...currentState,
+                ...action.extraData,
+            };
+            break;
+        case 'removeProperties':
+            currentState = { ...currentState };
+            for (const key of action.keysToRemove) {
+                delete currentState[key];
+            }
+            break;
+        default:
+            throw new Error(`Unknown action type: ${action.type}`);
```

Comment:
The `switch` statement is correctly used to handle different action types. However, to avoid repeating yourself, consider creating a shallow copy of the `currentState` outside of the `switch` cases, as the same operation is performed in multiple cases.